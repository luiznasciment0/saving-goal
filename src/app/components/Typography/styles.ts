import styled, { css } from 'styled-components'
import { TypographyProps } from '.'

export const sizeModifiers = {
  xsmall: () => css`
    font-size: 1.2rem;
    line-height: 1.6rem;
  `,
  small: () => css`
    font-size: 1.4rem;
    line-height: 2.1rem;
  `,
  medium: () => css`
    font-size: 1.6rem;
    line-height: 2rem;
  `,
  large: () => css`
    font-size: 1.8rem;
    line-height: 2.16rem;
  `,
  xlarge: () => css`
    font-size: 2rem;
    line-height: 2.4rem;
  `,
  xxlarge: () => css`
    font-size: 2.4rem;
    line-height: 2.88rem;
  `
}

const colorModifier = {
  primaryBlue: () => css`
    color: #1b31a8;
  `,
  secondaryBlue: () => css`
    color: #0079ff;
  `,
  primaryBlack: () => css`
    color: #1e2a32;
  `,
  secondaryBlack: () => css`
    color: #1c1e1f;
  `,
  primaryGray: () => css`
    color: #708797;
  `,
  secondaryGray: () => css`
    color: #4d6475;
  `
}

const fontFamilyModifier = {
  primary: () => css`
    font-family: 'Work Sans';
  `,
  secondary: () => css`
    font-family: 'Rubik';
  `
}

const weightModifier = {
  light: () => css`
    font-weight: 400;
  `,
  normal: () => css`
    font-weight: 500;
  `,
  bold: () => css`
    font-weight: 600;
  `
}

export const StyledTypography = styled.h1<TypographyProps>`
  ${({ color, size, family, weight }) => css`
    ${!!size && sizeModifiers[size]}
    ${!!family && fontFamilyModifier[family]}
    ${!!color && colorModifier[color]}
    ${!!weight && weightModifier[weight]}
  `}
`
