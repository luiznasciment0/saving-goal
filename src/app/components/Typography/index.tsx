import { StyledTypography } from './styles'

export type WeightProps = 'light' | 'normal' | 'bold'

export type SizeProps =
  | 'xsmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xlarge'
  | 'xxlarge'

export type ColorProps =
  | 'primaryBlue'
  | 'secondaryBlue'
  | 'primaryBlack'
  | 'secondaryBlack'
  | 'primaryGray'
  | 'secondaryGray'

export interface TypographyProps {
  children: React.ReactNode
  as?: React.ElementType
  family?: 'primary' | 'secondary'
  weight?: WeightProps
  size?: SizeProps
  color?: ColorProps
}

const Typography = ({
  children,
  as = 'h1',
  family = 'primary',
  weight = 'normal',
  size = 'medium',
  color = 'primaryBlue'
}: TypographyProps) => (
  <StyledTypography
    family={family}
    size={size}
    weight={weight}
    color={color}
    as={as}
  >
    {children}
  </StyledTypography>
)

export default Typography
