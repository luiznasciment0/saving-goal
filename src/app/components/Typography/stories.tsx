import { Story, Meta } from '@storybook/react/types-6-0'
import Typography from '.'

export default {
  title: 'Typography',
  component: Typography,
  argTypes: {
    children: {
      type: 'string'
    },
    tag: {
      type: 'string'
    },
    family: {
      type: 'string'
    },
    weight: {
      type: 'string'
    },
    size: {
      type: 'string'
    },
    color: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story = (args) => <Typography {...args}>Teste</Typography>

Default.args = {
  children: 'Teste'
}
