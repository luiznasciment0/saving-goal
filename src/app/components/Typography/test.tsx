import { render, screen } from '@testing-library/react'

import Typography from '.'

describe('<Typography />', () => {
  it('should render the medium size, normal weight, primaryBlue color and primary family by default', () => {
    const { container } = render(<Typography>Teste</Typography>)

    expect(screen.getByRole('heading', { name: /teste/i })).toHaveStyle({
      'font-size': '1.6rem',
      'line-height': '2rem',
      'font-weight': '500',
      'font-family': `'Work Sans'`,
      color: '#1b31a8'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the xsmall size`, () => {
    const { container } = render(<Typography size="xsmall">Teste 2</Typography>)

    expect(screen.getByRole('heading', { name: /teste 2/i })).toHaveStyle({
      'font-size': '1.2rem',
      'line-height': '1.6rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the small size`, () => {
    const { container } = render(<Typography size="small">Teste 3</Typography>)

    expect(screen.getByRole('heading', { name: /teste 3/i })).toHaveStyle({
      'font-size': '1.4rem',
      'line-height': '2.1rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the medium size`, () => {
    const { container } = render(<Typography size="medium">Teste 4</Typography>)

    expect(screen.getByRole('heading', { name: /teste 4/i })).toHaveStyle({
      'font-size': '1.6rem',
      'line-height': '2rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the large size`, () => {
    const { container } = render(<Typography size="large">Teste 5</Typography>)

    expect(screen.getByRole('heading', { name: /teste 5/i })).toHaveStyle({
      'font-size': '1.8rem',
      'line-height': '2.16rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the xlarge size`, () => {
    const { container } = render(<Typography size="xlarge">Teste 6</Typography>)

    expect(screen.getByRole('heading', { name: /teste 6/i })).toHaveStyle({
      'font-size': '2rem',
      'line-height': '2.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the xxlarge size`, () => {
    const { container } = render(
      <Typography size="xxlarge">Teste 7</Typography>
    )

    expect(screen.getByRole('heading', { name: /teste 7/i })).toHaveStyle({
      'font-size': '2.4rem',
      'line-height': '2.88rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the secondaryGray color`, () => {
    const { container } = render(
      <Typography color="secondaryGray">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#4d6475'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the primaryGray color`, () => {
    const { container } = render(
      <Typography color="primaryGray">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#708797'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the secondaryBlack color`, () => {
    const { container } = render(
      <Typography color="secondaryBlack">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#1c1e1f'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the primaryBlack color`, () => {
    const { container } = render(
      <Typography color="primaryBlack">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#1e2a32'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the secondaryBlue color`, () => {
    const { container } = render(
      <Typography color="secondaryBlue">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#0079ff'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the primaryBlue color`, () => {
    const { container } = render(
      <Typography color="primaryBlue">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      color: '#1b31a8'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the light weight`, () => {
    const { container } = render(<Typography weight="light">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '400'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the normal weight`, () => {
    const { container } = render(<Typography weight="normal">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '500'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the bold weight`, () => {
    const { container } = render(<Typography weight="bold">Hello</Typography>)

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-weight': '600'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the Rubik font-family`, () => {
    const { container } = render(
      <Typography family="secondary">Hello</Typography>
    )

    expect(screen.getByRole('heading', { name: /hello/i })).toHaveStyle({
      'font-family': `'Rubik'`
    })

    expect(container.firstChild).toMatchSnapshot()
  })
})
