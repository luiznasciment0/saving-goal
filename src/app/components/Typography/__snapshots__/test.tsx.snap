// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> should render the Rubik font-family 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Rubik';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the bold weight 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 600;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the large size 1`] = `
.c0 {
  font-size: 1.8rem;
  line-height: 2.16rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 5
</h1>
`

exports[`<Typography /> should render the light weight 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 400;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the medium size 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 4
</h1>
`

exports[
  `<Typography /> should render the medium size, normal weight, primaryBlue color and primary family by default 1`
] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste
</h1>
`

exports[`<Typography /> should render the normal weight 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the primaryBlack color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlack"
>
  Hello
</h1>
`

exports[`<Typography /> should render the primaryBlue color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the primaryGray color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #708797;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryGray"
>
  Hello
</h1>
`

exports[`<Typography /> should render the secondaryBlack color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #1c1e1f;
  font-weight: 500;
}

<h1
  class="c0"
  color="secondaryBlack"
>
  Hello
</h1>
`

exports[`<Typography /> should render the secondaryBlue color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #0079ff;
  font-weight: 500;
}

<h1
  class="c0"
  color="secondaryBlue"
>
  Hello
</h1>
`

exports[`<Typography /> should render the secondaryGray color 1`] = `
.c0 {
  font-size: 1.6rem;
  line-height: 2rem;
  font-family: 'Work Sans';
  color: #4d6475;
  font-weight: 500;
}

<h1
  class="c0"
  color="secondaryGray"
>
  Hello
</h1>
`

exports[`<Typography /> should render the small size 1`] = `
.c0 {
  font-size: 1.4rem;
  line-height: 2.1rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 3
</h1>
`

exports[`<Typography /> should render the xlarge size 1`] = `
.c0 {
  font-size: 2rem;
  line-height: 2.4rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 6
</h1>
`

exports[`<Typography /> should render the xsmall size 1`] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 2
</h1>
`

exports[`<Typography /> should render the xxlarge size 1`] = `
.c0 {
  font-size: 2.4rem;
  line-height: 2.88rem;
  font-family: 'Work Sans';
  color: #1b31a8;
  font-weight: 500;
}

<h1
  class="c0"
  color="primaryBlue"
>
  Teste 7
</h1>
`
